validator/src/tendermint.rs

```diff
@@ -330,134 +329,40 @@ where
-            };
-            macro_block.body = Some(body);
-
-            // Get a write transaction to the database, even if we don't intend to actually write
-            // anything out.
-            let mut txn = blockchain.write_transaction();
-
-            // Verify macro block state before committing accounts.
-            if let Err(error) = blockchain.verify_block_state_pre_commit(state, &block, &txn) {
-                debug!(%error, %block, "Tendermint - await_proposal: Invalid macro block state");
-                return Err(ProposalError::InvalidProposal);
-            }
-
-            // Update our blockchain state using the received proposal. If we can't update the state, we
-            // return a proposal timeout.
-            if let Err(error) = blockchain.commit_accounts(
-                state,
-                &block,
-                None,
-                &mut (&mut txn).into(),
-                &mut BlockLogger::empty(),
-            ) {
-                debug!(%error, %block, "Tendermint - await_proposal: Failed to commit accounts");
-                return Err(ProposalError::InvalidProposal);
-            }
-
-            // Check the validity of the block against our state. If it is invalid, we return a proposal
-            // timeout. This also returns the block body that matches the block header
-            // (assuming that the block is valid).
-            if let Err(error) = blockchain.verify_block_state_post_commit(state, &block, &txn) {
-                log::debug!(%error, %block, "Tendermint - await_proposal: Invalid block state");
-                return Err(ProposalError::InvalidProposal);
-            }
+        };
 
-            // Return the body of the macro block so it can be cached.
-            Ok(Body(block.unwrap_macro().body.unwrap()))
-        }
+        // Do the proposal verification.
+        blockchain
+            .verify_macro_block_proposal(
+                proposed_block,
+                proposal.message.round,
+                proposal.message.valid_round,
+                Self::hash_proposal(&proposal.message),
+                &proposal.signature.0,
+                signature_only,
+            )
+            .map(Body)
+            .map_err(|error| {
+                log::debug!(?error, ?proposal, "Proposal verification failed",);
+                // Special case for invalid signatures, as that means the proposer is not the signer.
+                // It is needed as the proposer if he produces a faulty block might be subjected to
+                // some sort of punishment or mitigation whereas with an invalid signature it would be
+                // the one relaying the message, who would potentially be subjected to protective measures.
+                if error == PushError::InvalidBlock(BlockError::InvalidJustification) {
+                    ProposalError::InvalidSignature
+                } else {
+                    ProposalError::InvalidProposal
+                }
```

Comment:
Just an idea, not mandatory: Maybe `PushError` is the wrong error type for `verify_macro_block_proposal`, maybe it could return a `ProposalError` directly?

Reply:
`ProposalError` is a fairly unspecific error type. I chose to only convert it here such that it can be logged before the conversion to hopefully help with debugging and for the, albeit highly unlikely chance, it would potentially see use in a different location. 

`ProposalError` is also an export of `nimiq-tendermint` which `nimiq-blockchain` currently only uses as a dev dependency, which I wanted to maintain.

Personally I would rather keep the `PushError` as it also is in line with the remaining functions present in `blockchain/src/blockchain/verify.rs` but changing it is not a big deal if it is considered better form.